zanata-war/src/main/java/org/zanata/rest/editor/service/SuggestionsService.java

```diff
@@ -0,0 +1,142 @@
+
+    @In("localeServiceImpl")
+    private LocaleService localeService;
+
+    @Override
+    public Response query(List<String> query, String sourceLocaleString, String transLocaleString, String searchTypeString) {
+
+        Either<SearchType, Response> searchTypeOrErrorResponse = getSearchType(searchTypeString);
+        if (searchTypeOrErrorResponse.isRight()) {
+            return searchTypeOrErrorResponse.right();
+        }
+        SearchType searchType = searchTypeOrErrorResponse.left();
+
+        Either<LocaleId, Response> sourceLocaleOrError = getSourceLocale(sourceLocaleString);
+        if (sourceLocaleOrError.isRight()) {
+            return sourceLocaleOrError.right();
+        }
+        LocaleId sourceLocale = sourceLocaleOrError.left();
+
+        Either<LocaleId, Response> transLocaleOrError = getTransLocale(transLocaleString);
+        if (transLocaleOrError.isRight()) {
+            return transLocaleOrError.right();
+        }
+        LocaleId transLocale = transLocaleOrError.left();
+
+        List<Suggestion> suggestions = transMemoryService.searchTransMemoryWithDetails(transLocale, sourceLocale,
+                new TransMemoryQuery(query, searchType));
+
+        // Wrap in generic entity to prevent type erasure, so that an
+        // appropriate MessageBodyReader can be used.
+        // see docs for GenericEntity
+        GenericEntity<List<Suggestion>> entity = new GenericEntity<List<Suggestion>>(suggestions) {};
+
+        return Response.ok(entity).build();
+    }
+
+    private Either<LocaleId, Response> getSourceLocale(String localeString) {
+        return getLocaleOrError(localeString, "Unrecognized source locale: \"%s\"");
+    }
+
+    private Either<LocaleId, Response> getTransLocale(String localeString) {
+        return getLocaleOrError(localeString, "Unrecognized translation locale: \"%s\"");
+    }
+
+    /**
+     * Try to get a valid locale for a given string. If the string does not match a supported
+     * locale, generate an error response with the given error message.
+     *
+     * @param localeString used to look up the locale
+     * @param errorFormat error to display. Include "%s" to substitute the locale string.
+     * @return a LocaleId if the given string matches one, otherwise an error response.
+     */
+    private Either<LocaleId, Response> getLocaleOrError(final String localeString, final String errorFormat) {
+        final String errorMessage = String.format(errorFormat, localeString);
+        final Response.ResponseBuilder errorResponse = Response.status(BAD_REQUEST).entity(errorMessage);
+
+        try {
+            @Nullable HLocale hLocale = localeService.getByLocaleId(localeString);
+            if (hLocale == null) {
+                return Either.right(errorResponse.build());
+            }
+            return Either.left(hLocale.getLocaleId());
+        } catch (IllegalArgumentException e) {
+            return Either.right(errorResponse.build());
```

Comment:
It's generally best to prevent exceptions from happening in the first place.  Catching exceptions without logging can lead to a lot of pain, and worth avoiding as a habit.

You could always extract the check from LocaleId's constructor to a public method, and call that first if you suspect an illegal locale code.  That's probably an improvement to LocaleId's API.